### ISSUE TITLE
Add JWT support

### ISSUE THREAD

**Issue by timkelty** [markhuot/craftql#89]
I have a project where I need to implement a GraphQL api, using JWT for access and authentication.

Still working out the best way to get this should work, but starting this ticket just to get some things down.

For my use-case, I'm looking at two types of access:

- "granted" access, e.g. a shared JWT token, generated from CraftQL (similar to how permission/access works now)
- "authenticated" access (e.g. for member access), where a request with user/pass would be posted to a Craft endpoint, returning a JWT to the client.
  - Subsequent requests to the server would include the JWT in the `Authentication: bearer` header
  - The token would be parsed on the server, determining what the access level was

In both cases, the body of the JWT would be the user info/access/permission levels.

For authenticated access, it seems the tricky part is syncing/refreshing the expiration of the token. It seems like you'd have to refresh the expiration of the token with every subsequent request somehow.

There are some low-level JWT methods here:
- https://github.com/flipboxfactory/craft-jwt

These articles have made the most sense to me:
- https://scaphold.io/community/questions/graphql-security-best-practices/
- https://scaphold.io/community/blog/authentication-in-graphql/

Would love any input from @nateiler as well!

**Comment by timkelty** [comment]
Thanks for jumping in @nateiler!

Some questions…

> be cautious as the JWT body is publicly viewable AND you should verify the permissions on the server side (to prevent tampering)

Wouldn't your JWT be encrypted with secret key, so you'd be able to trust it?

```
// header
{
  "alg": "HS256",
  "typ": "JWT"
}

// payload
{
  "sub": "1234567890",
  "name": "John Doe",
  "admin": true
}

// signature
HMACSHA256(
  base64UrlEncode(header) + "." +
  base64UrlEncode(payload),
  <secret_key>
)
```

**Comment by nateiler** [comment]
I was incorrect.  The body could be tampered with, but on the token verification side this SHOULD be caught via the signature verification process.  There a couple forms of JWT:    

A JWE type token would be encrypted and the the contents are unknown.  Contents are unknown to the end user

A JWS type token is not encrypted and the contents could be viewed.  It's up to consumer to verify the  signature which usually contains the body.

https://jwt.io/

**Comment by markhuot** [comment]
Thanks for all this info! I'm still a little hazy on the benefit of JWT over just a "token" in the traditional sense?

Right now: you generate a token in CraftQL. That token can have permissions applied so it can only query and not mutate, for example. Or, can only query one entry type not all entry types. You can have any many of these tokens as you'd like and they're all random so they shouldn't be "guessable."

JWT: would allow me to encode some information _within_ the token, but I'm not sure what information I would even need, other than the account information. But isn't that account information the token, as it's currently written?

So, what benefit would JWT provide? I could encrypt the query, itself, as it's sent over?

I see the advantages of JWT but I'm just now seeing how they could be practically included in to CraftQL, yet. I'd love to discuss this more though because it does seem like something I could add in without too much trouble.

**Comment by nateiler** [comment]
Mark, the point you made about issuing a token for non-mutation actions is pretty harmless.  While the tokens are random, and not easy to guess the outcome is security thru obscurity.  All you would need is some user to send a tweet with a token and then the entire world could use it and access the same data.  Rare, but it could happen.

A JWT token contains an identity in the body.  It's issued from your web app and (in this case) validated/verified by your web app.  If a security concern raises, you can address it on a per-use bases instead of potentially re-issuing a new token to all consumers.

Mutations should really be cautious of this as one tech savvy user could exploit this (and you wouldn't be able to pinpoint who did it).

In the case of CraftQL, the request would include a JWT authorization header along with the a standard GraphQL request/data.  On the server side, you would grab that Authorization header, validate the token and establish an identity.  From this point forward, all actions (service requests) are taken on behalf of the identity provided ... so `Craft::$app->getUsers()->getIdentity()->can('foo-bar')` is available, etc.

**Comment by timkelty** [comment]
> Under the current flow a user creates a token in the Craft UI and copies/pastes it. Is this okay for your specific use case or do you need a sign in route that a user can enter Craft credentials and be granted a JWT?

@markhuot right, in my case, I need to have the token returned from a sign-in.

**Comment by markhuot** [comment]
I wasn’t thinking of including anything important in the body of the token (since it's passed in the clear).

So, for the following example: “user logs in and is presented with a dashboard of their content” the flow would be,

1. user sees a login form
2. user enters username/pass
3. credentials are posted to CraftQL at `/api` via the GraphQL request below
4. CraftQL validates the user and generates a token for them, the token doesn’t contain any personally identifying information. It does not include their user id or their groups. The response data, however, does return that information you're looking for
5. front-end gets back the token and stores it in a cookie or local storage
7. front-end can use the rest of the response from `authorize` however it sees fit

The `authorize` request would look like this,

```graphql
{
  authorize(username: String!, password: String!) { #returns a User object
    token
    id
    groups {
      id
      name
    }
  }
}
```

**Comment by nateiler** [comment]
@timkelty Correct.  A token that is signed only, doesn't have an encrypted body.  The signature is technically hashed, not encrypted, using the algo identified in the header of the token.   Since you're issuing it to yourself, this doesn't really matter much.  

The signature will probably be handled via the JWT library.  You'll need to provide a secret.  Craft/Yii have a Security component which may help with this.

@markhuot I would say that CraftQL doesn't need to support JWT authorization natively.  There are numerous ways that one may want to authorize with your API.  My assumption is, all incoming API requests run through a single controller; if that's the case you could implement a behavior event (similar to https://github.com/craftcms/cms/blob/c0532b799d48e45b614f85466cc85db11bd54891/src/base/Component.php#L33) which would give devs the ability to get creative.

**Comment by davorpeic** [comment]
I'm open to beta test this.. :)

**Comment by markhuot** [comment]
It's happening (but not even close to done yet…).

https://github.com/markhuot/craftql/pull/103

![giphy](https://user-images.githubusercontent.com/48975/40858726-46d0b8b0-65ad-11e8-951a-effe268343f7.gif)

**Comment by markhuot** [comment]
I have a _very_ early pass of this over on the `dev-user-tokens` branch. If you have the time to pull it I'd be open to any critiques. The basic gist is,

First, ask for a token,

```graphql
{
  authorize(username:"foobar" password:"foobar") {
    token
    user {
      ...userFields
    }
  }
}
```

The `token` field will return the JWT token. Use that in your `Authorization: Bearer {token}` header and you'll be authenticated as the user who asked for the token.

In order to implement this I had to add CraftQL permissions to the user permission system. That means you'll need to check your user and for anyone other than an admin add the correct CraftQL permissions. That should look like this,

<img width="502" alt="screen shot 2018-06-12 at 1 41 30 pm" src="https://user-images.githubusercontent.com/48975/41307131-583b9d58-6e46-11e8-9e09-a739e7b90bc1.png">

Right now the user tokens do not expire. I'd like to offer that as a setting, but haven't gotten to that yet.

Please let me know if any of this seems usable for your use cases or if there's something missing.

**Comment by markhuot** [comment]
This is ready to merge. I'm going to do a bit more testing on it, but it'll probably be in `master` next week some time.

**Comment by timkelty** [comment]
I'll try and take a look this/next week too! Excited!

**Comment by markhuot** [comment]
Has anyone had a chance to test this out yet?

**Comment by timkelty** [comment]
@markhuot tragically, no. I SWEAR I'm getting to it early this week though 😀

**Comment by markhuot** [comment]
No worries! I just wanted to make sure I wasn't holding anyone up. I'm using this on `dev-user-tokens` for some upcoming work so it'll merge at some point either way.

**Comment by jan-thoma** [comment]
Regarding the expiry of tokens. I was working on an own solution but maybe switch to this project if JWT Tokens are fully implemented. Form my experience it could work something like this (i'currently using firebase/php-jwt):

- Add an expiry field to the JWT token.
- On every request decode the token it will fail if it's expired, abort with a 403
- If the token is valid issue a new one and send it via the auth header

**Comment by markhuot** [comment]
@jan-thoma, this is implemented _almost_ as you describe. Each JWT token has an `exp` field that contains the expiration timestamp. The duration is set via a `config/craftql.php` via `userTokenDuration`.

A request does not _automatically_ increase that expiration… although I like that idea. Currently you would need to do the following:

```graphql
query (username: String!, password: String!) {
  authorize(username:$username, password:$password) {
    token
  }
}
```

That would get you the initial token. Then to refresh it you'd need to,

```graphql
query(token: String!) {
  token: refresh(token:$token)
}
```

That would get you a new token with an updated `exp`.

What I like about this is you don't have to keep swapping out tokens on _every_ request. However, I get the feeling this isn't _the_ way to do JWT. Do you have any examples of where JWT is implemented and uses a constantly updating `exp`?

**Comment by jan-thoma** [comment]
This here was my approach, the code is heavy alpha stage but it might helps:

````
<?php
/**
 * Craft3 jwt plugin for Craft CMS 3.x
 *
 * Generates and validates JWT Tokens
 *
 * @link      https://t-k-f.ch
 * @copyright Copyright (c) 2018 jan.thoma@t-k-f.ch
 */

namespace tkf\craft3jwt\services;

use tkf\craft3jwt\Craft3Jwt;
use craft\helpers\DateTimeHelper;

use Craft;
use craft\base\Component;
use yii\web\HttpException;
use \Firebase\JWT\JWT as PhpJwt;

/**
 * @author    jan.thoma@t-k-f.ch
 * @package   Craft3Jwt
 * @since     1.0.0
 */
class Jwt extends Component
{
    // Private Properties
    // =========================================================================

    /**
     * @var boolean
     */
    private $authorized = false;

    /**
     * @var string
     */
    private $message = '';

    /**
     * @var number
     */
    private $status = 200;

    /**
     * @var object
     */
    private $token = null;

    /**
     * @var string
     */
    private $loginName = false;

    /**
     * @var string
     */
    private $password = false;

    /**
     * @var object
     */
    private $user = null;

    // Public Methods
    // =========================================================================

    /*
     * @return mixed
     */
    public function requestJwtToken()
    {
        $this->getOptionsRequest();

        try {
            $this->loginName = Craft::$app->request->getRequiredBodyParam('loginName');
            $this->password = Craft::$app->request->getRequiredBodyParam('password');
        }
        catch (\Exception $error)
        {
            return $this->errorToJson($error);
        }

        $this->validateUser();

        if (!$this->authorized)
        {
            return $this->errorToJson(new HttpException($this->status, $this->message));
        }

        return [
            'data' =>
            [
                'success' => true
            ]
        ];
    }

    /*
     * @return mixed
     */
    public function validateJwtToken()
    {
        $this->getOptionsRequest();

        try {
            $authorization = \Craft::$app->request->headers->get('authorization');
            $jwt = explode(' ', $authorization)[1];
        }
        catch (\Exception $e)
        {
            $this->abortRequest(400, 'authorization header missing or malformed');
        }

        try {
            $this->token = phpJwt::decode($jwt, \Craft::$app->config->general->securityKey, array('HS256'));
        }
        catch (\Exception $e)
        {
            $this->abortRequest(403, 'token invalid');
        }

        $this->validateUser();

        if (!$this->authorized)
        {
            $this->abortRequest($this->status, $this->message);
        }

        return $this->user;
    }

    /*
     * @return mixed
     */

    // Private Methods
    // =========================================================================

    /*
     * @return mixed
     */
    private function abortRequest ($status, $message)
    {
        throw new HttpException($status, $message);
    }

    /*
     * @return mixed
     */
    private function createJwtToken ()
    {
        $timestamp = DateTimeHelper::currentTimeStamp();
        $key = \Craft::$app->config->general->securityKey;
        $token = array(
            'iss' => "https://example.com",
            'aud' => "https://sub.example.com",
            'iat' => $timestamp,
            'exp' => $timestamp + $this->getSettings()['jwtExpiration'],
            'usr' => $this->user->uid
        );

        $jwt = phpJwt::encode($token, $key);

        \Craft::$app->response->headers->set('Authorization: Bearer', $jwt);
        \Craft::$app->response->headers->set('Access-Control-Expose-Headers', 'Authorization');
    }

    /*
     * @return mixed
     */
    private function errorToJson ($error)
    {
        \Craft::$app->response->setStatusCode($error->statusCode);

        return [
            'error' =>
            [
                'code' =>  $error instanceof HttpException ? $error->statusCode : $error->getCode(),
                'message' => $error->getMessage()
            ]
        ];
    }

    /*
     * @return mixed
     */
    private function getOptionsRequest ()
    {
        if(\Craft::$app->request->getMethod() === 'OPTIONS')
        {
            exit($this->setOptionsHeaders());
        }
    }

    /*
     * @return mixed
     */
    private function getSettings()
    {
        return Craft3jwt::$plugin->getSettings();
    }

    /*
     * @return mixed
     */
    private function setMessage ($authorized, $message)
    {
        $this->authorized = $authorized;
        $this->status = ($authorized) ? 200 : 403;
        $this->message = $message;
    }

    /*
     * @return mixed
     */
    private function setOptionsHeaders ()
    {
        header('Access-Control-Allow-Origin: *');
        header('Access-Control-Allow-Headers: Authorization');
        header('Access-Control-Allow-Methods: GET, POST');
    }

    /*
     * @return mixed
     */
    private function validateUser ()
    {
        if ($this->loginName)
        {
            $this->user =  Craft::$app->users->getUserByUsernameOrEmail($this->loginName);
        }

        if ($this->token)
        {
            $this->user =  \Craft::$app->users->getUserByUid($this->token->usr);
        }

        if (!$this->user)
        {
            $this->setMessage(false, 'invalid username or password');

            return;
        }

        if ($this->user->locked)
        {
            $this->setMessage(false, 'user locked for ' . DateTimeHelper::humanDurationFromInterval($this->user->remainingCooldownTime));

            return;
        }

        if ($this->user->suspended)
        {
            $this->setMessage(false, 'user suspended');

            return;
        }

        if ($this->password)
        {
            $authorized = $this->user->authenticate($this->password);

            $this->setMessage($authorized, ($authorized) ? 'success' : 'wrong username or password');

            if ($this->authorized)
            {
                $this->createJwtToken();
            }

            return;
        }

        if ($this->token && $this->user)
        {
            $this->authorized = true;
            $this->createJwtToken();

            return;
        }
    }
}
````

**Comment by markhuot** [comment]
Yup, seems similar to https://github.com/markhuot/craftql/blob/user-tokens/src/Types/Query.php#L362-L373.

Also with `Firebase\JWT\JWT` it has built in checking for expiration, which is great.

I'll look in to the rolling expiration times since it shouldn't affect backwards compatibility and removes the need to keep making a `refresh` query.

**Comment by jan-thoma** [comment]
It would be great then to have the option to create an application wide key. With the same scope options as an user has which not expires, to access everything that is considered public.

**Comment by markhuot** [comment]
This branch doesn't forego the existing Token system that is non-user-based. So, you should still be able to create a token and manage the permissions down to what you need. That token (as is currently the case) will never expire.

In short there will be two token types moving forward:

- user tokens: tied to user permissions and expire
- app tokens: not tied to a user and never expire

**Comment by markhuot** [comment]
I've added in support for rolling JWT tokens. Basically the initial request would still be the same,

```graphql
query (username: String!, password: String!) {
  authorize(username:$username, password:$password) {
    token
  }
}
```

That'll get you back a `token` string that you can store in your app. Send that back via the `Authorization` header, like so:

```
Authorization: Bearer {$token}
```

The change, is that when CraftQL responds it'll send an `Authorization` header back in the response with an updated token. The exact response will look like this,

```
HTTP/1.1 200 OK
Date: Wed, 18 Jul 2018 21:57:24 GMT
Server: Apache
X-Powered-By: Craft Commerce,Craft CMS
Authorization: TOKEN eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6IjQwIiwiZXhwIjoxNTMxOTY1NDQ0fQ.hvZE7Qz5jej6HTuhVrIFvj2v3W7epRW2nCVanvPw9OY
Allow: POST, GET
Connection: close
Transfer-Encoding: chunked
Content-Type: application/json; charset=UTF-8

{"data":{"entries":[{"id":41}]}}
```

I did some digging and don't really see a standard header name for CraftQL to respond with so `Authorization` seemed like the best fit. Has anyone used anything better? `X-Auth-Token` came up quite a few times too…

**Comment by jan-thoma** [comment]
### Following to [RFC 6750](https://tools.ietf.org/html/rfc6750) which describes 0Auth

When sending the access token in the "Authorization" request header
   field defined by HTTP/1.1 [RFC2617], the client uses the "Bearer"
   authentication scheme to transmit the access token.

   For example:

```
     GET /resource HTTP/1.1
     Host: server.example.com
     Authorization: Bearer mF_9.B5f-4.1JqM
```

So using the "Authorization: Bearer" in the request and the response should follow the specs in the closest manner.

**Comment by markhuot** [comment]
Thanks @jan-thoma. I updated this to use the `Authorization: Bearer {token}` format to better align with the spec.

However, I'm not seeing anything in [RFC 6750](
https://tools.ietf.org/html/rfc675) that indicates this should be the case for _responses_. The text you copied seems to be discussing the _request_ instead. Nonetheless, it seems okay to sync the two up.

**Comment by timkelty** [comment]
@markhuot finally digging into this, and it seems just what I was hoping for! 👏 

To clarify:

If you're sending all your requests with your updated token, you shouldn't need to manually call `refresh(token:$token)`, correct?

For my uses, for this to be useful, I need a bit more info in the body, namely which member groups the user is in.

Obviously different applications will need different jwt body specs…is it possible to expose this via event hook, so it can be customized?

**Comment by jan-thoma** [comment]
once you have the token you just can call your userdata via the api. i think you shouldn't expose as less as possible userdata in the token because anybody can decode the payload.

**Comment by markhuot** [comment]
@timkelty, I'm worried about using the JWT body for this since GraphQL actually supports this sort of multidimensional retrieval on any request. For example, if I added a `me` field that would return the user accessing data via token, you could do this,

```graphql
{
  entries: entries(limit: 5) {
    id
    title
    url
  }
  user: me {
    id
    groups {
      id
      name
    }
  }
}
```

Like you say, the benefit here is that you have more control over the data you're requesting and it's not _as_ public.

What do you think?

**Comment by timkelty** [comment]
@markhuot yep, that seems like a good solution to me.

**Comment by timkelty** [comment]
@markhuot been working on integrating this, and came across my first hurdle:

Apparently, to be able to inspect the `Authorization` header in the response, `Access-Control-Expose-Headers` needs to be set (this was news to me: https://stackoverflow.com/questions/28107459/in-the-http-cors-spec-whats-the-difference-between-allow-headers-and-expose-he)

Adding `$response->headers->add('Access-Control-Expose-Headers', 'Authorization');` seemed to work: https://github.com/markhuot/craftql/blob/master/src/Controllers/ApiController.php#L51

**Comment by timkelty** [comment]
Any update on this one?

I think the only remaining issues I'm having are:
- needing a `me`/`currentUser` field, as [mentioned above](https://github.com/markhuot/craftql/issues/89#issuecomment-412610847)
- Access-Control-Expose-Headers issue, [mentioned above](https://github.com/markhuot/craftql/issues/89#issuecomment-413676611)

**Comment by gertst** [comment]
Can we have an example on how to implement this?

**Comment by chrisrowe** [comment]
+1 for this. I'm holding off on until https://github.com/markhuot/craftql/pull/103 closes out

**Comment by u12206050** [comment]
I have created a PR into user-tokens to support individual entry type permissions.
#200
